File: tests/solutions_cpp/reorder_list.cpp
#include "json_helper.h"

class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head || !head->next) return;

        // Step 1: Find middle using slow/fast pointers
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // Step 2: Reverse second half
        ListNode* second = slow->next;
        slow->next = nullptr; // cut the list
        ListNode* prev = nullptr;
        while (second) {
            ListNode* tmp = second->next;
            second->next = prev;
            prev = second;
            second = tmp;
        }
        second = prev;

        // Step 3: Merge two halves
        ListNode* first = head;
        while (second) {
            ListNode* tmp1 = first->next;
            ListNode* tmp2 = second->next;
            first->next = second;
            second->next = tmp1;
            first = tmp1;
            second = tmp2;
        }
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto head = build_list(args[0]);
    sol.reorderList(head);
    cout << list_to_json(head) << endl;
    return 0;
}

File: tests/solutions_js/reorder_list.js
class Solution {
    reorderList(head) {
        if (!head || !head.next) return;

        // Step 1: Find middle using slow/fast pointers
        let slow = head;
        let fast = head;
        while (fast.next && fast.next.next) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // Step 2: Reverse second half
        let second = slow.next;
        slow.next = null;
        let prev = null;
        while (second) {
            const tmp = second.next;
            second.next = prev;
            prev = second;
            second = tmp;
        }
        second = prev;

        // Step 3: Merge two halves
        let first = head;
        while (second) {
            const tmp1 = first.next;
            const tmp2 = second.next;
            first.next = second;
            second.next = tmp1;
            first = tmp1;
            second = tmp2;
        }
    }
}
